contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,910 @@
+     *             instead of adding to an existing one.
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        bool _swapWeth,
+        uint256 _minTokenReceived
+    ) external nonReentrant onlyGovernorOrStrategist {
+        _rebalance(_amountToSwap, _minTokenReceived, _swapWeth);
+    }
+
+    /**
+     * @dev Remove almost all of the liqudity, rebalance the pool to the desired token split and
+     * deposit all of the liquidity.
+     */
+    // rebalance already has reentrancy check
+    // slither-disable-start reentrancy-no-eth
+    function _rebalance(
+        uint256 _amountToSwap,
+        uint256 _minTokenReceived,
+        bool _swapWeth
+    ) internal {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time there is no strategy's
+         * liquidity in the pool yet. The partial removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removeLiquidity(1e18);
+        }
+        // in some cases we will just want to add liquidity and not issue a swap to move the
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWeth);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        _checkForExpectedPoolPrice();
+
+        _addLiquidity();
+        _checkForExpectedPoolPrice();
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial or all liquidity from the pool.
+     * @param _liquidityToDecrease The amount of liquidity to remove expressed in 18 decimal point
+     */
+    function _removeLiquidity(uint256 _liquidityToDecrease) internal {
+        require(_liquidityToDecrease > 0, "Must remove some liquidity");
+
+        // unstake the position from the gauge
+        clGauge.withdraw(tokenId);
```

Comment:
🔴 This is withdrawn from the gauge but not replaced back on the gauge.

I think this means that after a withdraw, future withdraw() from the vault, and future rebalances would fail as well? Also that if we even withdraw 1 ETH, we stop earning?

Or am I missing something?

Reply:
So, this function gets called in three places: `rebalance()`, `withdraw()` and `withdrawAll()`. 

For `rebalance()`, it calls `addLiquidity` later, so it gets deposited to the gauge again.

For `withdrawAll()`, there shouldn't be anything to deposit back. So that's fine.

The problem seems to be with `withdraw()` where it unstakes to remove partial liquidity but doesn't deposit it back after that

Reply:
I would lean towards just keeping things consistent and always putting it back in the gauge when we are done doing something with the NFT. This way we never have to check where it is, and never have to if statement anywhere.

Reply:
Yes I agree, any future changes will minimise the possible errors if we remove/add after any liquidity change. I think the modifier is the most elegant way to facilitate the behaviour: 
https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/881a617a121896804e173c9bacccb8c4880dc25e

Reply:
Turns out you can not deposit into the Gauge if the token has no liqduitiy. Unfortunately this mean we need another bool to track whether the token has been deposited or not. Additional changes here: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/f551d7f37aa891880d3ed0893a7d894446680967